File: clients/python/src/vibium/async_api/browser.py
"""Async Browser class and launcher."""

from __future__ import annotations

from ..check import CheckResult, send_check
from ..run import RunResult, send_run

import json
import os
from typing import Any, Callable, Dict, List, Optional, Set, TYPE_CHECKING

from .page import Page
from .context import BrowserContext

if TYPE_CHECKING:
    from ..client import BiDiClient
    from ..binary import VibiumProcess


class Browser:
    """Async browser automation entry point."""

    def __init__(self, client: BiDiClient, process: Optional[VibiumProcess]) -> None:
        self._client = client
        self._process = process
        self._page_callbacks: List[Callable[[Page], None]] = []
        self._popup_callbacks: List[Callable[[Page], None]] = []
        self._seen_context_ids: Set[str] = set()

        # Listen for browsingContext.contextCreated events
        self._client.on_event(self._handle_event)

    def __repr__(self) -> str:
        return "Browser(connected=True)"

    def _handle_event(self, event: Dict[str, Any]) -> None:
        if event.get("method") != "browsingContext.contextCreated":
            return
        params = event.get("params", {})
        context_id = params.get("context")
        if not context_id or context_id in self._seen_context_ids:
            return
        self._seen_context_ids.add(context_id)
        callbacks = self._popup_callbacks if params.get("originalOpener") else self._page_callbacks
        if callbacks:
            page = Page(self._client, params["context"], params.get("userContext", "default"))
            for cb in callbacks:
                cb(page)

    async def __call__(self, goal: str, *, provider: Optional[str] = None, model: Optional[str] = None, base_url: Optional[str] = None, reasoning_effort: Optional[str] = None) -> RunResult:
        return await self.run(goal, provider=provider, model=model, base_url=base_url, reasoning_effort=reasoning_effort)

    async def run(self, goal: str, *, provider: Optional[str] = None, model: Optional[str] = None, base_url: Optional[str] = None, reasoning_effort: Optional[str] = None) -> RunResult:
        """Accomplish a goal in the live browser using the configured runtime."""
        return await send_run(self._client, goal, provider=provider, model=model, base_url=base_url, reasoning_effort=reasoning_effort)

    async def check(self, claim: str, *, record: Optional[str] = None, provider: Optional[str] = None, model: Optional[str] = None, base_url: Optional[str] = None, reasoning_effort: Optional[str] = None) -> CheckResult:
        """Independently verify live behavior, or inspect a read-only archive."""
        return await send_check(self._client, claim, record, provider=provider, model=model, base_url=base_url, reasoning_effort=reasoning_effort)


    async def page(self) -> Page:
        """Get the default page (first browsing context)."""
        result = await self._client.send("vibium:browser.page", {})
        return Page(self._client, result["context"], result.get("userContext", "default"))

    async def new_page(self) -> Page:
        """Create a new page (tab) in the default context."""
        result = await self._client.send("vibium:browser.newPage", {})
        return Page(self._client, result["context"], result.get("userContext", "default"))

    async def new_context(self) -> BrowserContext:
        """Create a new browser context (isolated, incognito-like)."""
        result = await self._client.send("vibium:browser.newContext", {})
        return BrowserContext(self._client, result["userContext"])

    async def pages(self) -> List[Page]:
        """Get all open pages."""
        result = await self._client.send("vibium:browser.pages", {})
        return [Page(self._client, p["context"], p.get("userContext", "default")) for p in result["pages"]]

    def on_page(self, callback: Callable[[Page], None]) -> None:
        """Register a callback for when a new page is created."""
        self._page_callbacks.append(callback)

    def on_popup(self, callback: Callable[[Page], None]) -> None:
        """Register a callback for when a popup is opened."""
        self._popup_callbacks.append(callback)

    def remove_all_listeners(self, event: Optional[str] = None) -> None:
        """Remove all listeners for 'page', 'popup', or all."""
        if not event or event == "page":
            self._page_callbacks.clear()
        if not event or event == "popup":
            self._popup_callbacks.clear()

    async def stop(self) -> None:
        """Stop the browser and clean up."""
        try:
            await self._client.send("vibium:browser.stop", {})
        except Exception:
            pass  # Browser or connection may already be closed
        await self._client.close()
        if self._process:
            await self._process.stop()


class _BrowserLauncher:
    """Module-level browser launcher object."""

    async def check(self, claim: str, *, record: str, executable_path: Optional[str] = None, provider: Optional[str] = None, model: Optional[str] = None, base_url: Optional[str] = None, reasoning_effort: Optional[str] = None) -> CheckResult:
        """Inspect an archive without installing or starting a browser."""
        from ..binary import VibiumProcess
        from ..client import BiDiClient
        if not record:
            raise ValueError("Standalone verification requires record")
        process = await VibiumProcess.start(no_browser=True, executable_path=executable_path)
        client = None
        try:
            client = await BiDiClient.connect(process)
            return await send_check(client, claim, record, provider=provider, model=model, base_url=base_url, reasoning_effort=reasoning_effort)
        finally:
            try:
                if client:
                    await client.close()
            finally:
                await process.stop()


    async def start(
        self,
        url: Optional[str] = None,
        *,
        engine: Optional[str] = None,
        channel: Optional[str] = None,
        headless: bool = False,
        headers: Optional[Dict[str, str]] = None,
        caps: Optional[Dict[str, Any]] = None,
        executable_path: Optional[str] = None,
    ) -> Browser:
        """Start a browser session.

        Args:
            url: Remote BiDi WebSocket URL, or an http(s) classic WebDriver
                endpoint (Selenium Grid, cloud grid). If not provided, checks
                VIBIUM_CONNECT_URL env var, then falls back to local launch.
            engine: Browser engine to launch: "chrome" (default) or "firefox"
                (local launch only).
            channel: Release channel of the engine to install and run, e.g.
                "beta". Currently honored by Firefox only (local launch only).
            headless: Run browser in headless mode (local launch only).
            headers: HTTP headers for remote connection (e.g. auth tokens).
            caps: Extra alwaysMatch capabilities for classic WebDriver
                endpoints (vendor-prefixed keys like vendor:options).
            executable_path: Path to vibium binary (default: auto-detect).
        """
        from ..binary import VibiumProcess
        from ..client import BiDiClient

        connect_url = url or os.environ.get("VIBIUM_CONNECT_URL")
        if connect_url:
            env_headers: Dict[str, str] = {}
            api_key = os.environ.get("VIBIUM_CONNECT_API_KEY")
            if api_key:
                env_headers["Authorization"] = f"Bearer {api_key}"
            merged = {**env_headers, **(headers or {})}
            # Raw JSON string either way — the binary validates it and owns
            # the error message, so no parsing here.
            caps_json = json.dumps(caps) if caps else os.environ.get("VIBIUM_CONNECT_CAPS")
            process = await VibiumProcess.start(
                connect_url=connect_url,
                connect_headers=merged or None,
                connect_caps=caps_json or None,
                executable_path=executable_path,
            )
        else:
            process = await VibiumProcess.start(
                headless=headless,
                engine=engine,
                channel=channel,
                executable_path=executable_path,
            )
        client = await BiDiClient.connect(process)
        return Browser(client, process)


browser = _BrowserLauncher()


class _EngineLauncher:
    """Named engine launcher, Playwright-style: firefox.start() is
    browser.start(engine="firefox")."""

    def __init__(self, engine: str) -> None:
        self._engine = engine

    async def start(
        self,
        url: Optional[str] = None,
        *,
        channel: Optional[str] = None,
        headless: bool = False,
        headers: Optional[Dict[str, str]] = None,
        executable_path: Optional[str] = None,
    ) -> Browser:
        return await browser.start(
            url,
            engine=self._engine,
            channel=channel,
            headless=headless,
            headers=headers,
            executable_path=executable_path,
        )


firefox = _EngineLauncher("firefox")
chrome = _EngineLauncher("chrome")

File: clients/javascript/src/browser.ts
import { callable } from './callable';
import { RunOptions, RunResult, sendRun } from './run';
import { CheckOptions, RecordedCheckOptions, CheckResult, sendCheck } from './check';
import { VibiumProcess } from './clicker';
import { BiDiClient, BiDiEvent } from './bidi';
import { Page } from './page';
import { BrowserContext } from './context';
import { debug, info } from './utils/debug';

const customInspect = Symbol.for('nodejs.util.inspect.custom');

export interface StartOptions {
  /** Browser engine to launch: 'chrome' (default) or 'firefox'. */
  engine?: 'chrome' | 'firefox';
  /**
   * Release channel of the engine to install and run, e.g. 'beta'.
   * Currently honored by Firefox only.
   */
  channel?: string;
  headless?: boolean;
  headers?: Record<string, string>;
  /** Extra alwaysMatch capabilities for classic WebDriver endpoints
   *  (cloud grids take their config this way, via vendor-prefixed capability keys). */
  caps?: Record<string, unknown>;
  executablePath?: string;
}

export interface Browser { (goal: string, options?: RunOptions): Promise<RunResult>; }

export class Browser {
  private client: BiDiClient;
  private process: VibiumProcess | null;
  private pageCallbacks: ((page: Page) => void)[] = [];
  private popupCallbacks: ((page: Page) => void)[] = [];

  constructor(client: BiDiClient, process: VibiumProcess | null) {
    this.client = client;
    this.process = process;

    // Listen for browsingContext.contextCreated events
    this.client.onEvent((event: BiDiEvent) => {
      if (event.method === 'browsingContext.contextCreated') {
        const params = event.params as {
          context: string;
          url: string;
          userContext?: string;
          originalOpener?: string;
        };

        // Only create a Page if there are callbacks to deliver it to.
        // Creating a Page registers event handlers (network, dialog) that
        // would otherwise leak and interfere with the real Page's handlers.
        const callbacks = params.originalOpener ? this.popupCallbacks : this.pageCallbacks;
        if (callbacks.length > 0) {
          const page = new Page(this.client, params.context, params.userContext || 'default');
          for (const cb of callbacks) {
            cb(page);
          }
        }
      }
    });
    return callable(this);
  }

  [customInspect](): string {
    return 'Browser { connected: true }';
  }

  /** Accomplish a live browser goal; provider settings are read in the runtime. */
  run(goal: string, options: RunOptions = {}): Promise<RunResult> {
    return sendRun(this.client, goal, options);
  }

  /** Independently verify this session, or inspect a saved archive. */
  check(claim: string, options: CheckOptions = {}): Promise<CheckResult> {
    return sendCheck(this.client, claim, options);
  }

  /** Get the default page (first browsing context). */
  async page(): Promise<Page> {
    const result = await this.client.send<{ context: string; userContext: string }>('vibium:browser.page', {});
    return new Page(this.client, result.context, result.userContext);
  }

  /** Create a new page (tab) in the default context. */
  async newPage(): Promise<Page> {
    const result = await this.client.send<{ context: string; userContext: string }>('vibium:browser.newPage', {});
    return new Page(this.client, result.context, result.userContext);
  }

  /** Create a new browser context (isolated, incognito-like). */
  async newContext(): Promise<BrowserContext> {
    const result = await this.client.send<{ userContext: string }>('vibium:browser.newContext', {});
    return new BrowserContext(this.client, result.userContext);
  }

  /** Get all open pages. */
  async pages(): Promise<Page[]> {
    const result = await this.client.send<{ pages: { context: string; url: string; userContext: string }[] }>('vibium:browser.pages', {});
    return result.pages.map(p => new Page(this.client, p.context, p.userContext));
  }

  /** Register a callback for when a new page is created (e.g. new tab). */
  onPage(callback: (page: Page) => void): void {
    this.pageCallbacks.push(callback);
  }

  /** Register a callback for when a popup is opened (window.open or target=_blank). */
  onPopup(callback: (page: Page) => void): void {
    this.popupCallbacks.push(callback);
  }

  /**
   * Remove all listeners for a given event, or all events if no event specified.
   * Supported events: 'page', 'popup'.
   */
  removeAllListeners(event?: 'page' | 'popup'): void {
    if (!event || event === 'page') {
      this.pageCallbacks = [];
    }
    if (!event || event === 'popup') {
      this.popupCallbacks = [];
    }
  }

  /** Stop the browser and clean up. */
  async stop(): Promise<void> {
    try {
      await this.client.send('vibium:browser.stop', {});
    } catch {
      // Browser or connection may already be closed
    }
    await this.client.close();
    if (this.process) {
      await this.process.stop();
    }
  }
}

function envHeaders(): Record<string, string> {
  const apiKey = process.env.VIBIUM_CONNECT_API_KEY;
  return apiKey ? { Authorization: `Bearer ${apiKey}` } : {};
}

export const browser = {
  /** Inspect a saved archive without installing or starting a browser. */
  async check(claim: string, options: RecordedCheckOptions): Promise<CheckResult> {
    if (!options?.record) throw new Error('Standalone verification requires record');
    const proc = await VibiumProcess.start({ noBrowser: true, executablePath: options.executablePath });
    let client: BiDiClient | undefined;
    try {
      client = BiDiClient.fromStreams(proc.stdin, proc.stdout, proc.preReadyLines);
      return await sendCheck(client, claim, options);
    } finally {
      try { await client?.close(); } finally { await proc.stop(); }
    }
  },
  async start(urlOrOptions?: string | StartOptions, options: StartOptions = {}): Promise<Browser> {
    let url: string | undefined;
    if (typeof urlOrOptions === 'object') {
      options = urlOrOptions;
      url = undefined;
    } else {
      url = urlOrOptions;
    }
    const connectURL = url || process.env.VIBIUM_CONNECT_URL;
    if (connectURL) {
      const headers = { ...envHeaders(), ...options.headers };
      // Raw JSON string either way — the binary validates it and owns the
      // error message, so no parsing here.
      const caps = options.caps ? JSON.stringify(options.caps) : process.env.VIBIUM_CONNECT_CAPS;
      debug('connecting to remote browser', { url: connectURL });

      const proc = await VibiumProcess.start({
        connectURL,
        connectHeaders: Object.keys(headers).length ? headers : undefined,
        connectCaps: caps || undefined,
        executablePath: options.executablePath,
      });
      debug('vibium started (connect mode)');

      const client = BiDiClient.fromStreams(
        proc.stdin,
        proc.stdout,
        proc.preReadyLines,
      );
      info('browser connected (pipe → remote)');

      return new Browser(client, proc);
    }

    const { engine, channel, headless = false, executablePath } = options;
    debug('launching browser', { engine, channel, headless, executablePath });

    const proc = await VibiumProcess.start({
      engine,
      channel,
      headless,
      executablePath,
    });
    debug('vibium started');

    const client = BiDiClient.fromStreams(
      proc.stdin,
      proc.stdout,
      proc.preReadyLines,
    );
    info('browser launched (pipe)');

    return new Browser(client, proc);
  },
};

/**
 * Named engine launchers, Playwright-style. `firefox.start()` is
 * `browser.start({ engine: 'firefox' })`; options are otherwise identical.
 */
export const firefox = {
  start(options: Omit<StartOptions, 'engine'> = {}): Promise<Browser> {
    return browser.start({ ...options, engine: 'firefox' });
  },
};

export const chrome = {
  start(options: Omit<StartOptions, 'engine'> = {}): Promise<Browser> {
    return browser.start({ ...options, engine: 'chrome' });
  },
};
